src/firebolt/utils/timeout_controller.py

```diff
@@ -0,0 +1,29 @@
+from time import time
+from typing import Optional
+
+from firebolt.utils.exception import QueryTimeoutError
+
+
+class TimeoutController:
+    """A class to control the timeout of a sequence of queries.
+    Used to check if the timeout has been reached and to calculate the remaining time.
+    """
+
+    timeout_error_cls = QueryTimeoutError
+
+    def __init__(self, timeout: Optional[float]):
+        self.timeout = timeout
+        self.start_time = time()
+
+    def check_timeout(self) -> None:
```

Comment:
nit: `raise_if_timeout` name is more fitting here as it's clearer that this raises an error and not just checks if we've reached the correct time.